### ISSUE TITLE
Question regarding pseudocode

### ISSUE THREAD

**Issue by NielsRogge** [conceptofmind/SemDeDup#1]
Hi,

Cool that you're reimplementing the paper :)

So I see your code is heavily inspired by the pseudocode they provided in the paper, and have a question regarding that.

So they state in their paper: 

> Within each cluster, we compute all pairwise cosine similarities and set a threshold cosine similarity above which data pairs are considered semantic duplicates. Finally, from each group of semantic duplicates within a cluster, we keep the image with the lowest cosine similarity to the cluster centroid and remove the rest.

So I understand this, however that doesn't seem to be equivalent to the pseudocode they provide. Namely, the pseudocode first sorts the images by distance to the cluster centroid (either most/least similar):

```
# Sort the cluster embeddings by the distance to the cluster
centroid .
cluster_i_embeddings = sort_by_distance_to_cluster_centroid(cluster_i_embeddings , descending = True )
```
and then takes the most similar image for each image:
```
M = torch . max( triu_sim_matrix , dim =0) [0]
```

, and finally removes all images whose most similar image falls below a certain threshold:

```
points_to_keep_from_cluster_i = M <= 1 - epsilon
```

In that pseudocode, I'm not seeing how those "groups of semantic duplicates" are created within a given cluster.

Do you happen to know that?

**Comment by conceptofmind** [comment]
Hi @NielsRogge ,

The current code in the repository is still a WIP and has been updated a few times (although not uploaded here) since. I agree that the pseudocode does not follow the outline. I emailed the paper author, Amro, who stated that he would do a review and provide feedback. I will have to email him again and see if he has checked it out.

It probably should be something like this instead:
```python
    # get upper triangular part of the matrix
    triu_sim_matrix = np.triu(pairwise_sim_matrix, k=1)

    # find lowest cosine similarity for each image
    most_similar_image_indices = np.argmin(triu_sim_matrix, axis=1)

    # Find image's similarity <= the threshold.
    points_to_keep_from_cluster_i = cluster_i_embeddings[
        triu_sim_matrix[np.arange(len(cluster_i_embeddings)), most_similar_image_indices] <= 1 - epsilon
    ]
```
I have not confirmed that it meets a certain level of correctness. I will need to further evaluate after the paper author's response but time has unfortunately been limited.

I am also only going to be using this for natural language.

Best,

Enrico

**Comment by NielsRogge** [comment]
Ok a review from the original authors would be very helpful! 

I'd like to apply this to filter image-text pairs.

**Comment by nehasrikn** [comment]
@NielsRogge @conceptofmind Just checking back in on this thread! Am also looking into this implementation, and was wondering if the authors had a chance to review and/or if @conceptofmind had any updates to this implementation!

**Comment by NielsRogge** [comment]
The best script I found to do k-means clustering on a large scale dataset is the one of Datacomp: https://github.com/mlfoundations/datacomp/blob/main/baselines/image_based_clustering.md. So for part 1 of the algorithm I'd use this one.

**Comment by NielsRogge** [comment]
I emailed the authors but go no response

**Comment by bharathc346** [comment]
The pseudo code seems to be erroneous from the paper. Here's what I believe it implies and what I think makes sense:

So they state in their paper:
> Within each cluster, we compute all pairwise cosine similarities and set a threshold cosine similarity above which data pairs are considered semantic duplicates. Finally, from each group of semantic duplicates within a cluster, we keep the image with the lowest cosine similarity to the cluster centroid and remove the rest.

The unclear part is
> Finally, from each group of semantic duplicates within a cluster, ..

To me each group of semantic duplicates is basically a connected component in a graph. So basically if we have A, B, C, D

````
# x indicates "duplicates" i.e cosine above 1 - epsilon

      A  B  C  D  E
A        x   
B             x  
C
D
E                       x
```

This basically indicates that (ABC) are one connected component or "group" comprised of pairs of semantic duplicates AB, BC and (DE) is one connected component that is comprised of the the semantic duplicate pair DE.

> we keep the image with the lowest cosine similarity to the cluster centroid and remove the rest.
Basically among ABC we take the one with lowest cosine similarity to the cluster centroid (motivation being you want data uniqueness) so  we take argmin_{A,B,C} cos_sim(·, cluster_centroid) for the ABC group and argmin_{D,E} cos_sim(·, cluster_centroid) for the DE group.

**Comment by conceptofmind** [comment]
I will have to contact the author again. I am looking to actively complete this implementation now.

**Comment by bharathc346** [comment]
https://github.com/taylorai/galactic/blob/main/src/galactic/semdedup.py

Here is an example of a similar implementation

**Comment by Isaachhh** [comment]
**The pseudocode in the paper is right.**

As bharathc346 said,

> To me each group of semantic duplicates is basically a weakly connected component in a graph. So basically if we have A, B, C, D
> 
> ```
> # x indicates "duplicates" i.e cosine above 1 - epsilon
> 
>    A B C D E
> A    x   
> B      x  
> C
> D           x
> E            
> ```
> 
> This basically indicates that (ABC) is one weakly connected component or "group" comprised of pairs of semantic duplicates AB, BC and (DE) is one weakly connected component that is comprised of the the semantic duplicate pair DE.
> 
> > we keep the image with the lowest cosine similarity to the cluster centroid and remove the rest.
> 
> Basically among ABC we take the one with lowest cosine similarity to the cluster centroid (motivation being you want data uniqueness) so we take argmin_{A,B,C} cos_sim(·, cluster_centroid) for the ABC group and argmin_{D,E} cos_sim(·, cluster_centroid) for the DE group.

We should keep only 1 in {ABC} and only 1 in {DE}.

Revist the pseudocode,
```
# Sort the cluster embeddings by the distance to the cluster centroid
cluster_i_embeddings = sort_by_distance_to_cluster_centroid(cluster_i_embeddings, descending = True )
M = torch.max( triu_sim_matrix, dim =0) [0]
points_to_keep_from_cluster_i = M <= 1 - epsilon
```

Sort the embeddings by the distance to the cluster centroid and then calculate upper triangular part of the cosine similarity matrix. 

So, for the j-th embedding (j-th column), only its cos-sim with i-th (i<j) embedding is considered. And that `M = torch.max( triu_sim_matrix, dim =0) [0]  > 1 - epsilon` means that, there exist i-th (i<j) embedding being too similar to j-th embedding and we should remove j-th embedding.

Then for all i-j pairs, we keep the smaller index and remove the bigger index. 

Therefore, in one weakly connected component, we keep the smallest index.

Noted that the smaller the index is, the bigger the distance to the centroid is.

So, in one weakly connected component, we keep the most far embedding, which is consistent with the paper.

Also, changing the order to smaller distance or randomness, we implement the most near or random de-duplicating.

**Comment by Isaachhh** [comment]
And welcome to follow our work of exploring pruning LAION-2B by SemDeDup and other strategies.

We supply the full-process code (easy to run on multiple nodes and GPUs) and an extensible framework.

See https://github.com/BAAI-DCAI/Dataset-Pruning/tree/main/LAION

**Comment by qmdnls** [comment]
The official implementation has now been released here:

https://github.com/facebookresearch/SemDeDup
